### ISSUE TITLE
Telegraf fails to reconnect to closed then reopened tcp connection

### ISSUE THREAD

**Issue by jedthe3rd** [influxdata/telegraf#4207]
### Relevant telegraf.conf:
https://gist.github.com/jedthe3rd/55ff4469030f23a3d8595ce8353dd300

### System info:

Telegraf v1.6.2, Windows Server 2012 R2, Receiving output is Logstash on a Redhat 7.4 box

### Steps to reproduce:

1. Start Logstash successfully.
2. Start telegraf.
3. Once telegraf is outputting to Logstash successfully restart logstash.
4. Look at the telegraf log and you will see it is failing to reconnect after the closed and reopened connection (due to restart) occurs.
5. Restart Telegraf to get the connection to work again.

### Expected behavior:
Telegraf would reconnect to Logstash after it finishes restarting and keep sending data, without having to restart Telegraf.

### Actual behavior:
Telegraf requires restart after Logstash is restarted. 

### Additional info:
Here is the error on Telegraf's side:

2018-05-29T21:36:10Z E! Error writing to output [socket_writer]: write tcp XXXXXXX->XXXXXX:5046: wsasend: An established connection was aborted by the software in your host machine.

No error on Logstash's side.

Similar problem described here: https://github.com/influxdata/telegraf/issues/4108

**Comment by danielnelson** [comment]
I added a bit of logging that should show if the connection is recreated or if it is reused (normally we reuse if the error "is temporary") 59f0a5354fd95ecb85c7807d84e807d29454d1b5.  Would it be possible for you to reproduce with [this build](https://12756-33258973-gh.circle-artifacts.com/0/build/windows/amd64/telegraf-1.7.0%7E59f0a535_windows_amd64.zip)?

**Comment by jedthe3rd** [comment]
Will do! Will report back soon.

**Comment by jedthe3rd** [comment]
So I ran it and I am not getting any new errors.

**Comment by jedthe3rd** [comment]
Here is the gist of the log: https://gist.github.com/jedthe3rd/cb17ef2ea9ddc92406352b58457bf86e
I let it run for a bit then stopped and started logstash.

**Comment by jedthe3rd** [comment]
I also tried to see if how quickly logstash restarts affects the result. I stopped logstash for ~10 min and then started again and telegraf responded the same as above.

**Comment by danielnelson** [comment]
Since the error does not start with `closing connection:`, then this error must be considered as a `Temporary()` error by the Go net package.

A little more research leads me to believe that this is the issue reported in https://github.com/golang/go/issues/24808.

**Comment by jedthe3rd** [comment]
What should be my next steps from here?

**Comment by danielnelson** [comment]
I think we will need to port the upstream fix into the Telegraf codebase.  Is this something you could work on?

**Comment by jedthe3rd** [comment]
I don't have enough experience in Go/Telegraf to do that properly. This isn't urgent atm.

**Comment by jedthe3rd** [comment]
@danielnelson I was looking further into this problem, wouldn't the issue fixed in [golang/go#24808](https://github.com/golang/go/issues/24808) be fixed for telegraf as well? Or does telegraf not automatically inherit changes from the latest net package? 
From what I can tell if those specific errors  (ECONNRESET/ECONNABORTED) get thrown it only classifies them as temporary for a specific case.

I think we need to find out what error is being specifically thrown. I will figure out how to get a dev environment running and see if I can't test this.

**Comment by jedthe3rd** [comment]
So after test building telegraf I now see that there is a specific version of the net package specified. Currently telegraf uses a337091b0525af65de94df2eb7e98bd9962dcbe2 but the fix is in 7f6105f138b3836e9ad85b8da26d44c742bf217b of the net package. So the fix would be to update the version that telegraf uses.

**Comment by danielnelson** [comment]
I looks to me that this fix is in the standard Go library, not the golang.org/x/net library.

**Comment by jedthe3rd** [comment]
So it is already applied? I am new to Go, so I am confused what needs to be done to telegraf to fix the problem.

**Comment by danielnelson** [comment]
I believe, though I don't know much about Go development process, that it will be in a future but undetermined Go release.

**Comment by jedthe3rd** [comment]
So the fix for now until they make a new release is to add a conditional check for that specific error.

**Comment by jedthe3rd** [comment]
So something along the lines of this:
https://gist.github.com/jedthe3rd/efa019762c9759ea4e54950f3a23bba8

**Comment by danielnelson** [comment]
Yes, this is more or less the fix.  We will need to wrap it so it will work on all platforms though.

**Comment by jedthe3rd** [comment]
Looks like they released the fix in the latest release, but it doesn't fix this issue. The code above also doesn't fix the issue. It closes the connection, but when it tries to reconnect it refuses to connect.

This is the error I get:
2018-06-11T23:49:10Z E! Error writing to output [socket_writer]: write tcp XXXXXXXX:XXXXX->XXXXXXXX:5046: wsasend: An existing connection was forcibly closed by the remote host.

**Comment by jedthe3rd** [comment]
Sorry, I miss understood what this [go/compare](https://github.com/golang/go/compare/go1.10.3...master ) was showing me. So no the fix isn't applied in the latest release. Still the code I posted doesn't fix the issue like we thought it would. I don't see that error. Which means that the error is happening before that. Probably in the sw.Conn check...

**Comment by jedthe3rd** [comment]
After further investigation it looks like wsasend is causing the problem. Just not sure why it is. However telegraf handles interfacing with that function (or the go functions that are in between) is causing problems. I feel like if a permanent error occurs that causes the err to return when it tries to connect, should it not close the connection? So [socket_writer.go/Write:120](https://gist.github.com/jedthe3rd/110c99f68bd8a6634ebff01a8f33b8e2) so we don't get the chance of persistent variables/connections.

**Comment by jedthe3rd** [comment]
@danielnelson I went in with the debugger to see why our 'fix' doesn't work and it looks like wsasend error isn't in that exception list. 
![image](https://user-images.githubusercontent.com/7817335/41383656-06a972fa-6f2f-11e8-80ea-445ac34b00a9.png)

We need to catch that error object. (under Err:). Any ideas how to reference that? wsasend isn't an accessible constant.


```
if err1, ok := err.(syscall.Errno); ok {
		log.Printf("Test2: %s", err1)
		err2 := sw.Close()
		sw.Conn = nil
		return fmt.Errorf("closing connection: %v : %v", err, err2)
}
```


That doesn't work which I hoped it would.

**Comment by danielnelson** [comment]
Thanks, I will incorporate this in a way that will work across platforms.

**Comment by apmattil** [comment]
I get the same issue with net.Conn.read().
I found same issue handled at  h2_bundle.go:http2isClosedConnError() about the same way as socket write above.

**Comment by danielnelson** [comment]
This issue is fixed in Go 1.11, which is used to build the official packages for Telegraf 1.9.0.

**Comment by dosmanak** [comment]
I had the problem with telegraf 1.10.1-n2. I changed the route for the remote output socket_writer and metrics were not received anymore even though the new route was working (I could open tcp connection using telnet)
